pkg/controller/scansettingbinding/scansettingbinding_controller.go

```diff
@@ -0,0 +1,515 @@
+	}
+
+	// finally retrieve the bundle that the profile points to
+	profReference.profileBundle, err = resolveProfileBundle(r, instance, &profReference, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	return &profReference, nil
+}
+
+func resolveProfile(r *ReconcileScanSettingBinding, instance *compliancev1alpha1.ScanSettingBinding, profReference *profileReference, logger logr.Logger) (*unstructured.Unstructured, error) {
+	return resolveTypedParent(r, instance, "Profile", profReference.tailoredProfile, logger)
+}
+
+func resolveProfileBundle(r *ReconcileScanSettingBinding, instance *compliancev1alpha1.ScanSettingBinding, profReference *profileReference, logger logr.Logger) (*unstructured.Unstructured, error) {
+	return resolveTypedParent(r, instance, "ProfileBundle", profReference.profile, logger)
+}
+
+func resolveTypedParent(r *ReconcileScanSettingBinding, instance *compliancev1alpha1.ScanSettingBinding, expectedKind string, child *unstructured.Unstructured, logger logr.Logger) (*unstructured.Unstructured, error) {
+	parentReference := ownerReferenceWithKind(child, expectedKind)
+	if parentReference == nil {
+		r.recorder.Eventf(
+			instance, corev1.EventTypeWarning, "BadReference",
+			"Couldn't find a %s owning %s %s", expectedKind, child.GetKind(), child.GetName(),
+		)
+		return nil, common.NewNonRetriableCtrlError("couldn't find an owner for %s %s owner", child.GetKind(), child.GetName())
+	}
+
+	logger.Info("Retrieving parent object",
+		"child.Kind", child.GetKind(), "child.Name", child.GetName(),
+		"parent.Name", parentReference.Name, "parent.Kind", expectedKind)
+
+	key := types.NamespacedName{Namespace: instance.Namespace, Name: parentReference.Name}
+	parentObj, err := getUnstructured(r, instance, key, parentReference.Kind, parentReference.APIVersion, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	if parentObj.GetKind() != expectedKind {
+		return nil, common.NewNonRetriableCtrlError("expected a %s, got %s", expectedKind, parentObj.GetKind())
+	}
+
+	return parentObj, nil
+
+}
+
+func ownerReferenceWithKind(object metav1.Object, kind string) *metav1.OwnerReference {
+	for _, ref := range object.GetOwnerReferences() {
+		if ref.Kind == kind {
+			return &ref
+		}
+	}
+
+	return nil
+}
+
+func getUnstructured(r *ReconcileScanSettingBinding, instance *compliancev1alpha1.ScanSettingBinding, key types.NamespacedName, kind, apiGroup string, logger logr.Logger) (*unstructured.Unstructured, error) {
+	logger.Info("Resolving object", "kind", kind, "api", apiGroup)
+
+	o := unstructured.Unstructured{}
+	o.SetAPIVersion(apiGroup)
+	o.SetKind(kind)
+
+	err := r.client.Get(context.TODO(), key, &o)
+	if errors.IsNotFound(err) {
+		return nil, common.NewRetriableCtrlErrorWithCustomHandler(func() (reconcile.Result, error) {
+			// This might be a temporary issue in the order the objects are being created
+			r.Eventf(
+				instance, corev1.EventTypeWarning, "NamedReferenceLookupError",
+				"NamedObjectReference %s not found", key,
```

Comment:
We probably want to add the kind to the event too, to help people figure out which object was the issue.

Reply:
done